src/toil_scripts/batch_alignment/bwa_alignment.py

```diff
@@ -1,223 +1,254 @@
-
-
-def run_bwa(job, inputs, ids):
+    :param dict ids: FileStore IDs for shared inputs
     """
-    Aligns two fastqs into a BAMFILE via BWA
-
-    :param JobFunctionWrappingJob job: Passed by Toil automatically
-    :param Namespace inputs: Input arguments (see main)
-    :param list ids: list of FileStore IDs (R1, R2, reference inputs)
-    """
-    work_dir = job.fileStore.getLocalTempDir()
-    file_names = ['r1.fq.gz', 'r2.fq.gz', 'ref.fa', 'ref.fa.amb', 'ref.fa.ann',
-                  'ref.fa.bwt', 'ref.fa.pac', 'ref.fa.sa', 'ref.fa.fai']
-    if inputs.alt:
-        file_names.append('ref.fa.alt')
-
-    for fileStoreID, name in zip(ids, file_names):
-        job.fileStore.readGlobalFile(fileStoreID, os.path.join(work_dir, name))
-
-    rg = ''
-    if inputs.rg_line:
-        rg = inputs.rg_line
+    uuid, urls = sample
+    r1_url, r2_url = urls if len(urls) == 2 else urls[0], None
+    job.fileStore.logToMaster('Downloaded sample: {0}\nStarting BWA Run: {0}'.format(uuid))
+    ids['r1'] = job.addChildJobFn(download_url_job, r1_url, s3_key_path=inputs.ssec, disk=inputs.file_size).rv()
+    if r2_url:
+        ids['r2'] = job.addChildJobFn(download_url_job, r2_url, s3_key_path=inputs.ssec, disk=inputs.file_size).rv()
     else:
-        rg = "@RG\\tID:{0}".format(inputs.uuid) # '\' character is escaped so bwakit gets individual '\' and 't' characters 
-        for tag, info in zip(['LB', 'PL', 'PU', 'SM'], ['library', 'platform', 'program_unit', 'uuid']):
-            if hasattr(inputs, info):
-                rg = rg + '\\t{0}:{1}'.format(tag, getattr(inputs, info))
-
-    # BWA Options
-    opt_args = []
-    if inputs.sort:
-        opt_args.append('-s')
-    if inputs.trim_adapters:
-        opt_args.append('-a')
-    # Call: bwakit
-    parameters = (['-t', str(inputs.cores),
-                   '-R', rg] +
-                  opt_args +
-                  ['-o', '/data/aligned',
-                   '/data/ref.fa',
-                   '/data/r1.fq.gz',
-                   '/data/r2.fq.gz'])
-    outputs = {'aligned.aln.bam': inputs.mock_bam}
-
-    docker_call(tool='quay.io/ucsc_cgl/bwakit:0.7.12--528bb9bf73099a31e74a7f5e6e3f2e0a41da486e',
-                parameters=parameters, inputs=file_names, outputs=outputs, work_dir=work_dir)
-
-    # BWA insists on adding an `*.aln.sam` suffix, so rename the output file
-    output_file = os.path.join(work_dir, '{}.bam'.format(inputs.uuid))
-    os.rename(os.path.join(work_dir, 'aligned.aln.bam'),
-              output_file)
-
-    # Either write file to local output directory or upload to S3 cloud storage
-    job.fileStore.logToMaster('Aligned sample: {}'.format(inputs.uuid))
-
-    if inputs.output_dir.startswith('s3://'):
-        s3am_upload(output_file, inputs.output_dir, s3_key_path=inputs.ssec)
+        ids['r2'] = None
+    # Define and wire job functions
+    bam_id = job.wrapJobFn(run_bwakit, inputs.cores, ids['r1'], ids['r2'], ids['ref'], ids['fai'], ids['amb'],
```

Comment:
> 19 parameters

That's a lot. Especially since they are passed by position. Good luck maintaining/refacorting that while not being on Adderall. It is ok to pass positional aka mandatory arguments by name. I'm aware that some of these parameters are sample-specific, but the non-sample-specific ones should be extracted from the config.

The point about the config is that it should be passed to every job, and every job extracts the information it needs from it. No unpacking or translation should happen. The config can also be modified, when a job imports a static input file, the file's URL is read from the config, the URL is imported, the config entry is modified by replacing the URL with the job store file ID and the modified config is returned.

This should work:

```py
config.ref = addChildJob( wrapJobFn( import_file_job( config.ref ) ) ).rv()
```